File: src/main/java/io/scriptor/impl/device/Memory.java
package io.scriptor.impl.device;

import io.scriptor.fdt.BuilderContext;
import io.scriptor.fdt.NodeBuilder;
import io.scriptor.machine.Device;
import io.scriptor.machine.IODevice;
import io.scriptor.machine.Machine;
import io.scriptor.util.Log;
import org.jetbrains.annotations.NotNull;

import java.io.PrintStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public final class Memory implements IODevice {

    private final Machine machine;
    private final long begin;
    private final long end;
    private final int capacity;
    private final boolean readonly;
    private final ByteBuffer buffer;

    public Memory(
            final @NotNull Machine machine,
            final long begin,
            final int capacity,
            final @NotNull ByteOrder order,
            final boolean readonly
    ) {
        this.machine = machine;
        this.begin = begin;
        this.end = begin + capacity;
        this.capacity = capacity;
        this.readonly = readonly;
        this.buffer = ByteBuffer.allocateDirect(capacity).order(order);
    }

    @Override
    public @NotNull Machine machine() {
        return machine;
    }

    @Override
    public void dump(final @NotNull PrintStream out) {
    }

    @Override
    public void reset() {
    }

    @Override
    public void build(final @NotNull BuilderContext<Device> context, final @NotNull NodeBuilder builder) {
        final var phandle = context.push(this);

        builder.name("memory@%x".formatted(begin))
               .prop(pb -> pb.name("phandle").data(phandle))
               .prop(pb -> pb.name("device_type").data("memory"))
               .prop(pb -> pb.name("reg").data(begin, end - begin));
    }

    @Override
    public long begin() {
        return begin;
    }

    @Override
    public long end() {
        return end;
    }

    @Override
    public long read(final int offset, final int size) {
        if (0 > offset || offset + size > capacity) {
            Log.error("invalid memory read offset=%x, size=%d", offset, size);
            return 0L;
        }

        return switch (size) {
            case 1 -> buffer.get(offset) & 0xFFL;
            case 2 -> buffer.getShort(offset) & 0xFFFFL;
            case 4 -> buffer.getInt(offset) & 0xFFFFFFFFL;
            case 8 -> buffer.getLong(offset);
            default -> {
                Log.error("invalid memory read offset=%x, size=%d", offset, size);
                yield 0L;
            }
        };
    }

    @Override
    public void write(final int offset, final int size, final long value) {
        if (readonly) {
            Log.error("invalid read-only memory write offset=%x, size=%d, value=%x", offset, size, value);
            return;
        }

        if (0 > offset || offset + size > capacity) {
            Log.error("invalid memory write offset=%x, size=%d, value=%x", offset, size, value);
            return;
        }

        switch (size) {
            case 1 -> buffer.put(offset, (byte) (value & 0xFFL));
            case 2 -> buffer.putShort(offset, (short) (value & 0xFFFFL));
            case 4 -> buffer.putInt(offset, (int) (value & 0xFFFFFFFFL));
            case 8 -> buffer.putLong(offset, value);
            default -> Log.error("invalid memory write offset=%x, size=%d, value=%x", offset, size, value);
        }
    }

    @Override
    public @NotNull String toString() {
        return "memory@%x".formatted(begin);
    }

    public void direct(final byte @NotNull [] data, final int offset, final boolean write) {
        if (0 > offset || offset + data.length > capacity) {
            Log.warn("direct read/write out of bounds: offset=%x, length=%d", offset, data.length);
            return;
        }

        if (write) {
            buffer.put(offset, data, 0, data.length);
        } else {
            buffer.get(offset, data, 0, data.length);
        }
    }

    public @NotNull ByteBuffer buffer() {
        return buffer;
    }
}

File: src/main/java/io/scriptor/impl/device/Memory.kt
package io.scriptor.impl.device

import io.scriptor.fdt.BuilderContext
import io.scriptor.fdt.NodeBuilder
import io.scriptor.machine.Device
import io.scriptor.machine.IODevice
import io.scriptor.machine.Machine
import io.scriptor.util.Log
import io.scriptor.util.Log.format
import java.io.FileOutputStream
import java.io.PrintStream
import java.nio.ByteBuffer

class Memory : IODevice {

    override val machine: Machine

    override val begin: ULong
    override val end: ULong

    private val capacity: UInt
    private val readonly: Boolean

    private val buffer: ByteBuffer

    constructor(machine: Machine, begin: ULong, capacity: UInt, readonly: Boolean) {
        this.machine = machine
        this.begin = begin
        this.capacity = capacity
        this.readonly = readonly

        this.end = begin + capacity
        this.buffer = ByteBuffer.allocateDirect(capacity.toInt()).order(machine.order)
    }

    override fun dump(out: PrintStream) {
        // ByteUtil.dump(out, buffer);
    }

    override fun reset() {
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun build(context: BuilderContext<Device>, builder: NodeBuilder) {
        val phandle = context.get(this)

        if (readonly) {
            builder
                .name(format("rom@%x", begin))
                .prop { it.name("phandle").data(phandle) }
                .prop { it.name("reg").data(begin, end - begin) }
                .prop { it.name("no-map").data() }
                .prop { it.name("readonly").data() }
        } else {
            builder
                .name(format("memory@%x", begin))
                .prop { it.name("phandle").data(phandle) }
                .prop { it.name("device_type").data("memory") }
                .prop { it.name("reg").data(begin, end - begin) }
        }
    }

    override fun read(offset: UInt, size: UInt): ULong? {
        if (offset + size > capacity) {
            Log.error("invalid memory read offset=%x, size=%d", offset, size)
            return null
        }

        return when (size) {
            1U -> buffer.get(offset.toInt()).toUByte().toULong()
            2U -> buffer.getShort(offset.toInt()).toUShort().toULong()
            4U -> buffer.getInt(offset.toInt()).toUInt().toULong()
            8U -> buffer.getLong(offset.toInt()).toULong()
            else -> {
                Log.error("invalid memory read offset=%x, size=%d", offset, size)
                null
            }
        }
    }

    override fun write(offset: UInt, size: UInt, value: ULong): Boolean {
        if (readonly) {
            Log.error("invalid memory write offset=%x, size=%d, value=%x, marked as read-only", offset, size, value)
            return false
        }

        if (offset + size > capacity) {
            Log.error("invalid memory write offset=%x, size=%d, value=%x", offset, size, value)
            return false
        }

        when (size) {
            1U -> buffer.put(offset.toInt(), value.toByte())
            2U -> buffer.putShort(offset.toInt(), value.toShort())
            4U -> buffer.putInt(offset.toInt(), value.toInt())
            8U -> buffer.putLong(offset.toInt(), value.toLong())
            else -> {
                Log.error("invalid memory write offset=%x, size=%d, value=%x", offset, size, value)
                return false
            }
        }

        return true
    }

    override fun toString(): String {
        return if (readonly) format("rom@%x", begin) else format("memory@%x", begin)
    }

    fun direct(data: ByteArray, offset: UInt, write: Boolean): Boolean {
        if (offset > capacity || offset + data.size.toUInt() > capacity) {
            Log.warn("direct read/write out of bounds: offset=%x, length=%d", offset, data.size)
            return false
        }

        if (write) {
            buffer.put(offset.toInt(), data, 0, data.size)
        } else {
            buffer.get(offset.toInt(), data, 0, data.size)
        }

        return true
    }

    fun buffer(): ByteBuffer {
        return buffer
    }

    fun export(filename: String) {
        FileOutputStream(filename).use { stream ->
            stream.channel.write(buffer)
        }
    }
}
